Add an interface to mark the solution as invalid / out-of-bounds

MengnanLi91 opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
If this PR implements an enhancement that does not have an existing issue, please add the following:
-->
ref #22814 Adding new feature to test whether the material property is valid

## Reason
<!--Why do you need this feature or what is the enhancement?-->
The feature is requested to determine whether the solution is valid at the time of convergence.

## Design
<!--A concise description (design) of the enhancement.-->
The function can be called from materials to state the converged solution is invalid. 
A bool function is added with a flag. The default value of the flag is false meaning solution is valid. The flag will flip every time the function is called. The final value will be the state of the converged solution.

## Impact
<!--Will the enhancement change existing APIs or add something new?-->
The new feature can state the converged solution is invalid.

<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

MengnanLi91 committed:
Solution Invalid refs #22814

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1255307/) on 5c29c3e wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22884/site)

This comment will be updated on new commits.

friedmud commented:
I should mention that this is not the _final_ solution for this issue... but it is _a_ solution that will close the ticket.

moosebuild commented:
Job [Examples and tutorials](https://civet.inl.gov/job/1255665/) on 7d628fa : invalidated by @cticenhour

Unrelated failure

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1255690/) on 7d628fa wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>af38e9</th>
      <th colspan=4>#22884 7d628f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.47%</td>
      <td>84.47%</td>
      <td>+0.00%</td>
      <td>90.91%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>80918</td>
      <td>80927</td>
      <td>+9</td>
      <td>10</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14872</td>
      <td>14873</td>
      <td>+1</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22884/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22884/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22884/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud reviewed (changes requested):
PR looks pretty good!
Some suggestions

While it s part of the solution to tag the solution as invalid, it s only on the final residual evaluation that we should do that. The Newton solve often visits numerous invalid states before converging on a valid one

So the 'invalid' status needs to be reset right before this final evaluation.

GiudGiud commented:
You ll want to rebase on upstream/devel instead of merging btw

friedmud commented:
@GiudGiud the flag resets before every residual evaluation... so the state of the flag at the end of the solve will represent what happened with the final solution.

GiudGiud commented:
Umm I must have missed that. Can one of you please point me to the line which does that?

friedmud commented:
@GiudGiud - hah!  You're right... the reset didn't make it into the PR.  It existed... but didn't make it into the commits!

GiudGiud commented on framework/src/systems/NonlinearSystem.C:
```diff
@@ -341,6 +341,12 @@ NonlinearSystem::converged()
 
+  if (_solution_invalid)
+  {
+    mooseWarning("The solution is not converged due to the solution being invalid.");
```
do you think we could cover this with just a set of names with an insert

the set would be reset at the same time as the boolean for invalid

MengnanLi91 replied:
I agree. This is the part I am working on. It will be added in later version of this feature.

MengnanLi91 commented on framework/include/interfaces/SolutionInvalidInterface.h:
```diff
@@ -0,0 +1,36 @@
+class FEProblemBase;
+
+/**
+ * An inteface for the _console for outputting to the Console object
```
This is updated.

MengnanLi91 commented on framework/include/interfaces/SolutionInvalidInterface.h:
```diff
@@ -0,0 +1,36 @@
+class SolutionInvalidInterface
+{
+public:
+  /**
+   * A class for providing a helper stream object for writting message to
+   * all the Output objects.
+   */
+  SolutionInvalidInterface(MooseObject * moose_object);
+  void solutionInvalid();
+
+private:
+  /// An instance of helper class to write streams to the Console objects
+  FEProblemBase * _si_fe_problem;
```
This is updated.

MengnanLi91 commented on framework/include/systems/SystemBase.h:
```diff
@@ -904,6 +904,14 @@ class SystemBase : public libMesh::ParallelObject, public ConsoleStreamInterface
+  /**
+   * Whether or not the converged solution is valid
+   */
+
```
This is updated.

MengnanLi91 commented on framework/include/systems/SystemBase.h:
```diff
@@ -904,6 +904,14 @@ class SystemBase : public libMesh::ParallelObject, public ConsoleStreamInterface
+   * Whether or not the converged solution is valid
+   */
+
+  bool solutionInvalid() const { return _solution_invalid; }
```
This function is often used with if condition(e.g. if(solutionInvalid) do something;), I feel statement without "is" is less confusing. What do you think?

GiudGiud replied:
seems fine now with the other renamed

MengnanLi91 commented on framework/include/systems/SystemBase.h:
```diff
@@ -904,6 +904,14 @@ class SystemBase : public libMesh::ParallelObject, public ConsoleStreamInterface
+
+  bool solutionInvalid() const { return _solution_invalid; }
+
+  void solutionInvalid(bool solution_invalid) { _solution_invalid = solution_invalid; }
```
Good point. This is updated.

MengnanLi91 commented on test/include/materials/NonsafeMaterial.h:
```diff
@@ -0,0 +1,32 @@
+#include "MaterialProperty.h"
+
+/**
+ * Simple material to test vector parameter range checking.
```
This is updated

MengnanLi91 commented on test/tests/materials/solution_invalid/tests:
```diff
@@ -0,0 +1,10 @@
+    expect_err = 'The solution is not converged due to the solution being invalid.'	
+    design = 'syntax/Materials/index.md'
+    issues = '#22814'
+    requirement = 'The system shall not allow a converged solution with an invalid computation.'
```
This is updated.

MengnanLi91 committed:
Updates to address reviewer's comments #22884

1. docstrings are updated.
2. reset issue is fixed.
3. namechange is done.

style change for Precheck test.

Apply suggestions from code review

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

Fix reset mechanism #22884

style changes for precheck test

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1259667/) on d9e05dc wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22884/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22884/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 85dd294ac64fd4d355dc774c1f6896960f59b4c2`

GiudGiud commented on framework/include/interfaces/SolutionInvalidInterface.h:
```diff
@@ -0,0 +1,35 @@
+class FEProblemBase;
+
+/**
+ * An inteface for communicate solution invalid state to FEProblemBase
```
```suggestion
 * An interface to communicate a solution invalid state to FEProblemBase
```

GiudGiud commented on framework/include/interfaces/SolutionInvalidInterface.h:
```diff
@@ -0,0 +1,35 @@
+  void setSolutionInvalid(bool solution_invalid);
+
+private:
+  /// An instance of helper class to communicate to FEProblemBase
```
```suggestion
  /// A pointer to the FEProblem
```

GiudGiud commented:
Another thing that will help a lot to deploy this everywhere is an option to bypass it, either on a Problem scale or for each material. 
‘basically a dont_set_invalid parameter but with a better name.

MengnanLi91 commented:
> 

Good point! I'll add it to the PR I am working on along with object tracking.

friedmud commented:
@GiudGiud - I think this is good to go.  We are already working on a follow-up PR where we are adding a lot of features.  Let's get this one in and close the "safety" ticket so @cticenhour can breathe easily!

GiudGiud reviewed (dismissed):
still no reset mechanism.

This is not usable as is. No app is going to want to invalidate their solution with no recourse (the boolean I asked for) and no information about what happened? (the list of objects involved, reminder that the warnings are being silenced)

Also the test is crashing on the warning, not on the return false on accepting the solution. It's not just missing a line, it's not testing what we want.

GiudGiud commented:
@cticenhour please review for concurrence

cticenhour commented:
OK @GiudGiud - will chime in this morning in between meetings.

friedmud commented:
Bah - I thought it had been fixed up.  My bad.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1261567/) on 95c6381 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22884/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22884/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 82cb1fbdfc8d797688fe2ab4796e5de33fb143c6`

cticenhour reviewed (dismissed):
I think this is the bare minimum required to satisfy the original issue now that the reset mechanism was just put in place. I have a few suggestions to make code comments and other quality-of-life stuff better. 

I agree with @GiudGiud that some "opt out" scenario should be available to those who want to implement this in their object and not use it for special cases - this is a vital piece of the feature set to me for it to be useful for apps. I disagree that it *needs* to be in this PR, but that disagreement is contingent on the ETA for the follow-up. When is this expected, and it is already being worked on in another branch?

cticenhour commented:
Note for all on the current set of test results so far - package builds will be failing until we get a next --> devel merge.

milljm commented:
You can also rebase on Next if you wish to not wait...

MengnanLi91 committed:
Allow overriding invalid solution check

MengnanLi91 committed:
Add doco for solution invalid and add newsletter blurb refs #22814

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1262074/) on 5cf1a95 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/22884/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/22884/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 82cb1fbdfc8d797688fe2ab4796e5de33fb143c6`

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 8d84cd9f3cffe24b781a005acb4c149767fcb56c
+Subproject commit 889a6e9e9bf06bdbe39220a528fc8f905b9c06df
```
**Caution!** This contains a submodule update

cticenhour replied:
This regression on the `large_media` submodule status needs to be removed from the commit history. Seems it was accidentally added.

cticenhour replied:
Resolved in d549645

loganharbour commented on framework/include/interfaces/SolutionInvalidInterface.h:
```diff
@@ -0,0 +1,35 @@
+
+private:
+  /// A pointer to the FEProblem
+  FEProblemBase * _si_fe_problem;
```
This should be a reference, not a pointer

friedmud commented:
@GiudGiud @cticenhour 

The branch is all rebased and squashed.  It has doco for the new interface and adds to the newsletter.  We added the ability to turn this check off.

As for why there isn't more information coming out: it's because it's not as straightforward as you think.  You cannot simply pass a string in to print out or add to a std::set or whatever - that is way too slow.  Think of the case where you needed to call this function on thousands (or millions) of quadrature points... it would _crawl_ if you were doing string operations.  What's needed is to have a registration system (similar to PerfGraph sections) so that you are just keeping track of counts of unique IDs.

That's what @MengnanLi91 is implementing now.  She's already a long way down the path due to picking up a generalized registration that @loganharbour had made a while ago.  The new system will (efficiently) keep track of how many times each piece of code set the solution to be invalid... and _when_... and be able to print it all out.

She's already a long way down the path - so I'm hopeful we will have it in by the end of next week.

THAT is why I don't want to put a ton of work into this PR - what is coming is WAY better... but this is good enough for now.

![d08b19c28aa886d6f7cf314fa32532b7](https://user-images.githubusercontent.com/870705/206801909-9b3e7bea-2cd7-4b51-b0f1-eb78a371572f.jpg)

cticenhour commented:
@friedmud I see your point - and really like the idea you are broadly proposing compared to this one. I am willing to approve this pending @loganharbour's comment being addressed (`_si_fe_problem` really should be a reference, not a pointer, but I missed that earlier).

GiudGiud reviewed (commented):
same approval as Casey

if you assume it would be done poorly, then yes it would be slow.
It s easy to imagine a workable solution that is not slow: keep track of the status with a boolean on each material, then do the "string" operation only when outputting the error message.

MengnanLi91 commented:
> 

LazyCoupleable.h use pointer while MortarConsumerInterface.h use reference for FEProblemBase. I am confused now...

loganharbour commented:
> > 
> 
> @cticenhour @loganharbour LazyCoupleable.h use pointer while MortarConsumerInterface.h use reference for FEProblemBase. I am confused now...

`LazyCouplable` was developed... lazily :-)

In general - you should only need to use a pointer if there is a valid reason for it to be null. In this case, we get an address to the problem on construction and it's always available. Therefore, a reference is more appropriate. Only use a pointer if a reference won't work for the situation.

This'll happen in time unfortunately - that is - finding old code where poor development choices were made. We don't have the resources to rewrite the code base with modern standards, but we do have the time to make sure that _new_ code is up to higher standards.

MengnanLi91 commented:
> > > 
> > 
> > 
> > @cticenhour @loganharbour LazyCoupleable.h use pointer while MortarConsumerInterface.h use reference for FEProblemBase. I am confused now...
> 
> `LazyCouplable` was developed... lazily :-)
> 
> In general - you should only need to use a pointer if there is a valid reason for it to be null. In this case, we get an address to the problem on construction and it's always available. Therefore, a reference is more appropriate. Only use a pointer if a reference won't work for the situation.
> 
> This'll happen in time unfortunately - that is - finding old code where poor development choices were made. We don't have the resources to rewrite the code base with higher standards, but we do have the time to make sure that _new_ code is up to higher standards.

Got it. I'll change it to a reference

cticenhour commented:
@loganharbour beat me to it while I was still typing, so I will delete and chime in with only this :) 

`LazyCoupleable` was expected not to have an address for `FEProblem` until after it was constructed, so a pointer was used so that it could be null until it was set.

MengnanLi91 committed:
Change the pointer to FEProblem with a reference

friedmud commented:
@loganharbour @cticenhour I disagree about the pointer business BTW.  The actual rule in MOOSE is that you try to avoid changing a pointer to a reference or a reference to a pointer.  If you receive something as a pointer then store/return it as one.  If you receive something as a reference then store/return it as one.

In this case the pointer is being received from the parameters as a pointer and there is nothing wrong with storing it as a pointer... especially since it's a private member of a rather small class.

We really need to loosen up on this kind of thing.  Let's debate stuff that matters, like public interfaces that users will see...

@GiudGiud , this ultimately won't be just for Materials.  An object might (probably) will have multiple different correlations that could be out of bounds.  You don't only want to know which object - you also want to know which correlation and get a reasonable print out about that particular correlation and what could be wrong.  Doing that requires a whole system for registering each place where `setSolutionInvalid()` is called with a good message to go along with it.

loganharbour commented:
> @loganharbour @cticenhour I disagree about the pointer business BTW. The actual rule in MOOSE is that you try to avoid changing a pointer to a reference or a reference to a pointer. 

This doesn’t make a shred of sense. We even check if the pointer is valid… there’s no reason to use a pointer in the slightest

GiudGiud commented:
Thank you for the patch @MengnanLi91 !

Congratulations on your first PR merged to MOOSE!

friedmud commented:
@GiudGiud - thanks for the merge!

lindsayad commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -1614,6 +1614,11 @@ class FEProblemBase : public SubProblem, public Restartable
+  /**
+   * Whether or not we are allowing invalid solutions
+   */
+  bool allowInvalidSolution() { return _allow_invalid_solution; }
```
Can be `const`

MengnanLi91 replied:
I'll update this in the next version of the feature.

lindsayad replied:
I'm very impressed with our quickly you're contributing @MengnanLi91!

lindsayad replied:
It must be the NCSU power 😄

MengnanLi91 replied:
Go wolfpack! I also got a lot of help from our awesome team to get my first PR done : )
![image](https://user-images.githubusercontent.com/118846840/207951373-eb93f8ba-f30f-4e9b-a839-50504cff38f6.png)

lindsayad commented:
I do think that in general it communicates a lot more safety to a developer when they are dealing with a reference instead of a pointer. When I get a pointer from an API my first inkling it to `mooseAssert(pointer, "This better not be null as I'm about to use it!");`

MengnanLi91 commented:
@GiudGiud I am working on improving the feature as we discussed. Below is the output looks like. Does it meet your need? Or is there any other features I should add?
<img width="528" alt="Screenshot 2022-12-15 at 3 08 18 PM" src="https://user-images.githubusercontent.com/118846840/207991660-1a090cf1-19fd-4a51-bfd9-bb0ae47b5529.png">
<img width="495" alt="Screenshot 2022-12-15 at 5 07 43 PM" src="https://user-images.githubusercontent.com/118846840/207992561-5c1e1ccc-c9f4-4191-853b-52d05a53e8a2.png">

GiudGiud commented:
We should get feedback from the analysis groups on this at some point

I think we would want to show messages instead of sections? 
a computeQpProperties could trigger solutionInvalid in multiple locations

MengnanLi91 commented:
> We should get feedback from the analysis groups on this at some point
> 
> I think we would want to show messages instead of sections? a computeQpProperties could trigger solutionInvalid in multiple locations

Below are the current version I have. What do you think? 
1. When there is solution invalid warning detected but allow_invalid_solution is true, the output is
![image](https://user-images.githubusercontent.com/118846840/208550865-84c7c62a-6ede-4273-9234-5495dc73dd99.png)
2. When there is solution invalid warning detected and not allowed, the output is
![image](https://user-images.githubusercontent.com/118846840/208550247-c6b5d4ce-780f-4341-9393-eddd9e74a66e.png)
3. When there is no solution invalid warning detected or no section registered for solution invalid check, the output is
![image](https://user-images.githubusercontent.com/118846840/208550340-ddf59ec8-54a0-4cae-8f41-fc0f7468822e.png)

GiudGiud commented:
Looks good.
So the 'Total iterations' it's the number of occurences? Over all linear, nonlinear and timesteps?
Or the number of iterations that presented the issue

MengnanLi91 commented:
> Looks good.
> 
> So the 'Total iterations' it's the number of occurences? Over all linear, nonlinear and timesteps?
> 
> Or the number of iterations that presented the issue

The total iterations count all warnings over all linear, nonlinear iterations and time steps while the last iteration will be reset for each iteration. Is it what we want to have? If this is okay,I could push this version for code review.

GiudGiud commented:
This total is somewhat obscure for users though. Users dont know that we take a residual evaluation step at the beginning and end of every time step for example.
It's a sum of residual and jacobian loops too, and that mapping to linear and non linear is somewhat obscure to users. 

I would say only give the final iteration # of occurrences
